src/Shared/Dictionary/AdaptiveCapacityDictionary.cs

```diff
@@ -0,0 +1,752 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.AspNetCore.Internal.Dictionary
+{
+    /// <summary>
+    /// An <see cref="IDictionary{String, Object}"/> type to hold a small amount of items (4 or less in the common case).
+    /// </summary>
+    internal class AdaptiveCapacityDictionary<TKey, TValue> : IDictionary<TKey, TValue>, IReadOnlyDictionary<TKey, TValue> where TKey : notnull
+    {
+        // Threshold for size of array to use.
+        private static readonly int DefaultArrayThreshold = 4;
+
+        internal KeyValuePair<TKey, TValue>[] _arrayStorage;
+        private int _count;
+        internal Dictionary<TKey, TValue>? _dictionaryStorage;
+        private IEqualityComparer<TKey> _comparer;
+
+        /// <summary>
+        /// Creates a new <see cref="AdaptiveCapacityDictionary{TKey, TValue}"/> from the provided array.
+        /// The new instance will take ownership of the array, and may mutate it.
+        /// </summary>
+        /// <param name="items">The items array.</param>
+        /// <param name="comparer">Equality comparison.</param>
+        /// <returns>A new <see cref="AdaptiveCapacityDictionary{TKey, TValue}"/>.</returns>
+        public static AdaptiveCapacityDictionary<TKey, TValue> FromArray(KeyValuePair<TKey, TValue>[] items, IEqualityComparer<TKey>? comparer = null)
+        {
+            if (items == null)
+            {
+                throw new ArgumentNullException(nameof(items));
+            }
+
+            comparer = comparer ?? EqualityComparer<TKey>.Default;
+
+            if (items.Length > DefaultArrayThreshold)
+            {
+                // Use dictionary for large arrays.
+                return new AdaptiveCapacityDictionary<TKey, TValue>(items, capacity: items.Length,comparer);
+            }
+
+            // We need to compress the array by removing non-contiguous items. We
```

Comment:
I think this behavior is routing specific. I don't know if we care in the general case.